lib/client/keystore.go

```diff
@@ -12,255 +12,261 @@ distributed under the License is distributed on an "AS IS" BASIS,
-func (a *LocalKeyAgent) CheckHostSignature(hostId string, remote net.Addr, key ssh.PublicKey) error {
-	cert, ok := key.(*ssh.Certificate)
-	if !ok {
-		return trace.Errorf("expected certificate")
-	}
-
-	bk, err := boltbk.New(filepath.Join(a.KeyDir, HostSignersFilename))
-	if err != nil {
-		return trace.Wrap(nil)
+// GetKeys returns all user session keys stored in the store
+func (fs *FSLocalKeyStore) GetKeys() (keys []Key, err error) {
+	dirPath := filepath.Join(fs.KeyDir, sessionKeyDir)
+	if !isDir(dirPath) {
+		return make([]Key, 0), nil
 	}
-	defer bk.Close()
-	ca := local.NewCAService(bk)
-
-	cas, err := ca.GetCertAuthorities(services.HostCA, false)
+	dirEntries, err := ioutil.ReadDir(dirPath)
 	if err != nil {
-		return trace.Wrap(err)
+		return nil, trace.Wrap(err)
 	}
-
-	for i := range cas {
-		checkers, err := cas[i].Checkers()
-		if err != nil {
-			return trace.Wrap(err)
+	for _, fi := range dirEntries {
+		if !fi.IsDir() {
+			continue
 		}
-		for _, checker := range checkers {
-			if sshutils.KeysEqual(cert.SignatureKey, checker) {
-				return nil
+		k, err := fs.GetKey(fi.Name())
+		if err != nil {
+			// if a key is reported as 'not found' it's probably because it expired
+			if !trace.IsNotFound(err) {
+				return nil, trace.Wrap(err)
+			} else {
+				continue
 			}
 		}
+		keys = append(keys, *k)
 	}
-	return trace.Errorf("no matching authority found")
+	return keys, nil
 }
 
-// getKeyes returns a list of local keys agents can use to authenticate
-func (a *LocalKeyAgent) GetKeys() ([]agent.AddedKey, error) {
-	existingKeys, err := a.loadAllKeys()
+// AddKey adds a new key to the session store. If a key for the host is already
+// stored, overwrites it.
+func (fs *FSLocalKeyStore) AddKey(host string, key *Key) error {
+	dirPath, err := fs.dirFor(host)
 	if err != nil {
-		return nil, trace.Wrap(err)
+		return trace.Wrap(err)
 	}
-	addedKeys := make([]agent.AddedKey, len(existingKeys))
-	for i, key := range existingKeys {
-		pcert, _, _, _, err := ssh.ParseAuthorizedKey(key.Cert)
+	writeBytes := func(fname string, data []byte) error {
+		fp := filepath.Join(dirPath, fname)
+		err := ioutil.WriteFile(fp, data, 0640)
 		if err != nil {
-			return nil, trace.Wrap(err)
+			log.Error(err)
 		}
-		pk, err := ssh.ParseRawPrivateKey(key.Priv)
-		if err != nil {
-			return nil, trace.Wrap(err)
-		}
-		addedKey := agent.AddedKey{
-			PrivateKey:       pk,
-			Certificate:      pcert.(*ssh.Certificate),
-			Comment:          "",
-			LifetimeSecs:     0,
-			ConfirmBeforeUse: false,
-		}
-		addedKeys[i] = addedKey
+		return err
 	}
-	return addedKeys, nil
+	if err = writeBytes(fileNameCert, key.Cert); err != nil {
+		return trace.Wrap(err)
+	}
+	if err = writeBytes(fileNamePub, key.Pub); err != nil {
+		return trace.Wrap(err)
+	}
+	if err = writeBytes(fileNameKey, key.Priv); err != nil {
+		return trace.Wrap(err)
+	}
+	ttl, _ := key.Deadline.UTC().MarshalJSON()
+	if err = writeBytes(fileNameTTL, ttl); err != nil {
```

Comment:
key TTL is encoded in the certificate associated with it, so why not simply get it from there